### PR TITLE
Refactor Promtail's `client.Manager` to do multi client instantiation

### PR DESCRIPTION
**Pull request by thepalbi**
**What this PR does / why we need it**:
This PR refactors Promtails `client.Manager` to include both WAL-enabled client, and WAL disabled. This will make the transition easier towards the agent codebase, only needing to instantiate one primitive to support both WAL enabled/disabled clients under the `loki.write` component. Also, re-uses a bunch of logic shared by both.

**Which issue(s) this PR fixes**:
Part of https://github.com/grafana/loki/issues/8197

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

**Review by cstyan — COMMENTED** [review]
Most LGTM, some minor comment changes and also I think the noop notifier needs to be easier to read at a glance. 

Lets also confirm with the agent team that, via the refactor you have here, the interface for promtail write clients would fit into their write component they way they want it to.

File: clients/pkg/promtail/client/manager_test.go [hunk 1]
```diff
@@ -136,3 +168,156 @@ func TestManager_WriteAndReadEntriesFromWAL(t *testing.T) {
...
+		}
+	}()
+
+	defer func() {
+		manager.Stop()
+		closeServer.Close()
+	}()
+
+	var testLabels = model.LabelSet{
+		"pizza-flavour": "fugazzeta",
```
[the remark this answers is not in the document]
Comment (thepalbi) [reply clients/pkg/promtail/client/manager_test.go:198]: Let me introduce you to Buenos Aires's pizza https://es.wikipedia.org/wiki/Fugazzetta

File: clients/pkg/promtail/client/manager.go [hunk 2]
```diff
@@ -76,36 +83,44 @@ func NewManager(
...
+		manager.startWithConsume()
+	} else {
+		manager.name = "multi"
+		manager.startWithForward()
+	}
 	return manager, nil
 }
 
-func (m *Manager) start() {
+// startWithConsume starts the main manager routine, which reads and discards entries from the exposed channel.
```
[the remark this answers is not in the document]
Comment (thepalbi) [reply clients/pkg/promtail/client/manager.go:122]: Added!

**Comment by thepalbi** [comment]
@cstyan addressed all comments, and simplified the nil notifer.

As for the the agent, their entry point is [here](https://github.com/grafana/agent/blob/main/component/loki/write/write.go#L119). Basically, they are using the same remote-write client in Loki, and doing on their own what the MultiClient does. By using the manager, we'll be able to abstract away this logic and use WAL.

As for using this refactor on their side, I asked them about how the bump the dependencies and they mentioned [this](https://raintank-corp.slack.com/archives/C02GSU8SHBN/p1689104519733849?thread_ts=1689102180.231199&cid=C02GSU8SHBN). ATM they are using a quite-old release of Loki, so it will require a bit of work copying the code + replacing some types with their own. I plan to do this effort myself, so I'll start drafting the changes with my local loki copy.

**Comment by thepalbi** [comment]
@cstyan submitted a draft PR forklifting the promtail codebase as in this pr https://github.com/grafana/agent/pull/4483

**Review by cstyan — APPROVED** [review]
Tested locally with two clients configured (not using WAL), still works :+1:

**Pull request merged**